### PR TITLE
tests: Enable automatic detection of undefined names in Python tests scripts. Remove wildcard imports.

### PR DESCRIPTION
**Pull request by practicalswift**
Enable automatic detection of undefined names in Python tests scripts. Remove wildcard imports.

Wildcard imports make it unclear which names are present in the namespace, confusing both readers and many automated tools.

An additional benefit of not using wildcard imports in tests scripts is that readers of a test script then can infer the rough testing scope just by looking at the imports.

Before this commit:

```
$ contrib/devtools/lint-python.sh | head -10
./test/functional/feature_rbf.py:8:1: F403 'from test_framework.util import *' used; unable to detect undefined names
./test/functional/feature_rbf.py:9:1: F403 'from test_framework.script import *' used; unable to detect undefined names
./test/functional/feature_rbf.py:10:1: F403 'from test_framework.mininode import *' used; unable to detect undefined names
./test/functional/feature_rbf.py:15:12: F405 bytes_to_hex_str may be undefined, or defined from star imports: test_framework.mininode, test_framework.script, test_framework.util
./test/functional/feature_rbf.py:17:58: F405 CScript may be undefined, or defined from star imports: test_framework.mininode, test_framework.script, test_framework.util
./test/functional/feature_rbf.py:25:13: F405 COIN may be undefined, or defined from star imports: test_framework.mininode, test_framework.script, test_framework.util
./test/functional/feature_rbf.py:26:31: F405 satoshi_round may be undefined, or defined from star imports: test_framework.mininode, test_framework.script, test_framework.util
./test/functional/feature_rbf.py:26:60: F405 COIN may be undefined, or defined from star imports: test_framework.mininode, test_framework.script, test_framework.util
./test/functional/feature_rbf.py:30:41: F405 satoshi_round may be undefined, or defined from star imports: test_framework.mininode, test_framework.script, test_framework.util
./test/functional/feature_rbf.py:30:68: F405 COIN may be undefined, or defined from star imports: test_framework.mininode, test_framework.script, test_framework.util
$
```

After this commit: 

```
$ contrib/devtools/lint-python.sh | head -10
$
```

### CODE REVIEW

**Comment by practicalswift** [comment]
@jnewbery Thanks for the concept ACK! One difference with this PR compared to the previous ones is that this makes sure no wildcard imports are re-introduced in the future thanks to the Travis check. So this should hopefully be the very last time we have to think about wildcard imports :-)

**Comment by MarcoFalke** [comment]
Just as a note: As a work around when writing tests, I replace the import list with a wildcard-star and then run some de-wildcard python script before submitting the pull request.

**Comment by practicalswift** [comment]
@MarcoFalke That's a convenient workflow! What tool do you use? [`dewildcard`](https://github.com/quentinsf/dewildcard)?

**Comment by JeremyRubin** [comment]
@practicalswift  I think my comment https://github.com/bitcoin/bitcoin/pull/9876#issuecomment-283165452 still stands, but I have no strong opinion on this.

**Comment by practicalswift** [comment]
Rebased!

**Comment by kallewoof** [comment]
Concept ACK, obviously.

**Comment by practicalswift** [comment]
Rebased!

**Comment by sipa** [comment]
This seems like a sensible policy, though I leave the decision up to people who more often work on the Python code.

**Comment by laanwj** [comment]
I don't care enough to NACK this, though it annoys me slightly if things that are rejected come up again and again.

**Comment by sdaftuar** [comment]
> though it annoys me slightly if things that are rejected come up again and again.

I feel the same way, and I'm still opposed to this idea -- I don't think we should be increasing the effort required for getting tests written and merged.  

My impression lately (which is admittedly completely unscientific so maybe others have a different impression) is that PRs that include tests seem to have a higher burden than those that don't, because it adds more surface area for review and potential for rebase (and it's easy to nit someone's test code).  I think this leads to too few tests in our project's PRs, as it is -- I don't think we should make this any worse by throwing up more hurdles.

But I don't know, there's lots of PRs like this one that I think are a waste of our time, but since this particular topic keeps coming up, if people want to spend their time on these kinds of things, I won't stand in the way...

**Comment by kallewoof** [comment]
> I feel the same way, and I'm still opposed to this idea -- I don't think we should be increasing the effort required for getting tests written and merged.

It is not an increase in effort to throw an error.
That's like saying our compilers should throw less errors cause the programmers get grumpy. Do you want more programmers who need a more gentle compiler?

To reiterate the comparison, if a contributor gives up *because of automated errors* they will probably give up halfway through review anyway. Why should we care so much about the imaginary lazy developer?

**Comment by practicalswift** [comment]
I see two possible states of the world:

* **State 1 – Travis does not check wildcard imports**: A Python PR introducing wildcard imports is opened. During the review process human reviewers (such as perhaps me, @kallewoof, @jnewbery, @JeremyRubin and @jtimon - who have historically voiced their concern about wildcard imports) will point out that these are problematic and discouraged. The human reviewer will explain that wildcard imports (`from <module> import *`) should be avoided, as they make it unclear which names are present in the namespace, confusing both readers and many automated tools. When looking at a specific introduction of a specific wildcard import in a specific individual case no one will argue against such a change (as no one would suggest going from explicit import to wildcard import in a specific case) and the consensus would thus land in fixing the wildcard import. The PR author switches to using explicit imports. The normal review process takes place focusing on non-trivial/non-mechanical stuff.

* **State 2 – Travis does check wildcard imports** A Python PR introducing wildcard imports is opened. Travis complains within one minute. The PR author switches to using explicit imports. The normal review process takes place focusing on non-trivial/non-mechanical stuff.

Energy used by the different participants:
* Energy used by PR author assuming state 1: Discussing review (`D`). Adjusting code (`A`).
* Energy used by PR author assuming state 2: Adjusting code (`A`).
* Energy used by `n` human reviewers assuming state 1: Reviewing (`n * R`).
* Energy used by `n` human reviewers assuming state 2: None (`0`).

Net effects:
* Energy used assuming state 1: `D + A + n * R`
* Energy used assuming state 2: `A`

Conclusion: I can see how status quo bias have put us in state 1, but what are compelling arguments for sticking with state 1 when it seems like everyone (PR author and human reviewers) is better off in state 2?

**Comment by ken2812221** [comment]
Concept ACK

**Comment by sdaftuar** [comment]
> During the review process human reviewers (such as perhaps me, @kallewoof, @jnewbery, @JeremyRubin and @jtimon - who have historically voiced their concern about wildcard imports) will point out that wildcard imports are problematic and discouraged.

I will refrain from debating particular style points (so whatever you guys want to do, go nuts).  But as a matter of process, I don't think style nits for things that are not part of the project's style guide are appropriate PR comments.  If you'd like to enforce a project-wide policy, I think you should start a discussion about that in a PR that would update the style guide.  Otherwise we should expect people with different personal style preferences to object for things that seem arbitrary.

Do we even have python style guidelines anywhere in our tree, apart from this linter in contrib/?  If we're going to impose rules, we should document them so contributors know what is expected.

**Comment by jamesob** [comment]
Concept ACK. 

Agree with @sdaftuar's point that it isn't fair to make style comments on PRs where the cited rule isn't present in a linter or styleguide. IMO "no wildcard imports" is inherent in "follow PEP8." 

Allowing wildcard imports degrades development experience in Python significantly because it reduces the effectiveness of in-editor linters like flake8. Once you allow wildcard imports, you can no longer detect legitimately undefined symbols. I think a change like this will make test development a bit more pleasant.

**Comment by practicalswift** [comment]
Rebased!

**Comment by Empact** [comment]
utACK 6185a96 this seems like a good improvement and low-risk to me

**Comment by ken2812221** [comment]
utACK 6185a96

**Comment by practicalswift** [comment]
Rebased!

File: test/functional/wallet_txn_doublespend.py [hunk 1]
```diff
@@ -3,6 +3,7 @@
 # Distributed under the MIT software license, see the accompanying
 # file COPYING or http://www.opensource.org/licenses/mit-license.php.
 """Test the wallet accounts properly when there is a double-spend conflict."""
+
```
Comment (Empact) [inline test/functional/wallet_txn_doublespend.py]: nit: unrelated whitespace change

File: test/functional/wallet_basic.py [hunk 2]
```diff
@@ -3,6 +3,7 @@
 # Distributed under the MIT software license, see the accompanying
 # file COPYING or http://www.opensource.org/licenses/mit-license.php.
 """Test the wallet."""
+
```
Comment (Empact) [inline test/functional/wallet_basic.py]: nit: unrelated whitespace change

File: test/functional/rpc_fundrawtransaction.py [hunk 3]
```diff
@@ -17,7 +17,6 @@
     find_vout_for_address,
 )
 
-
```
Comment (Empact) [inline test/functional/rpc_fundrawtransaction.py]: nit: unrelated whitespace change

**Comment by Empact** [comment]
Could avoid touching three of these files to keep the history clean.

**Comment by practicalswift** [comment]
@Empact Thanks for reviewing. Feedback addressed! Please re-review.

File: test/functional/p2p_segwit.py [hunk 4]
```diff
@@ -82,6 +82,7 @@
     sync_mempools,
 )
 
+
```
Comment (Empact) [inline test/functional/p2p_segwit.py]: One more

**Comment by practicalswift** [comment]
@Empact Please re-review :-)

**Comment by practicalswift** [comment]
Rebased!

**Comment by practicalswift** [comment]
Rebased! :-)

**Review by conscott — COMMENTED** [review]
Test ACK ea71fac1f228ec5525084b82c9c2c08a3517c692

Not to get pedantic on this debate, but I totally agree with @jamesob when he says
> Allowing wildcard imports degrades development experience in Python significantly because it reduces the effectiveness of in-editor linters like flake8. Once you allow wildcard imports, you can no longer detect legitimately undefined symbols. I think a change like this will make test development a bit more pleasant.

I would suspect most other Python programmers  also have these "in-editor linters" setup and have encountered this problem as well. This change will significantly reduce the amount of "lint barf" seen when saving test files with wildcards, and allow uncovering real undefined variables. 

If I see the false-positives like:
```
F999 <variable> may be undefined, or defined from star imports: test_framework.blocktools, test_framework.util
```
I am likely to ignore _everything_, whereas fixing the wildcard imports lets me see through the noise.

**Comment by ken2812221** [comment]
utACK ea71fac1f228ec5525084b82c9c2c08a3517c692

**Comment by practicalswift** [comment]
Rebased! Please re-review :-)

**Comment by laanwj** [comment]
On one hand I'm annoyed that apparently, insisting on doing something often enough gets it merged because the people that disagree will stop bothering to complain.

One the other hand, fine, this specific case is not a big deal.

**Comment by MarcoFalke** [comment]
It wasn't that only a small group of people insisted on doing it, but really a different person each time (see #9876 by @jnewbery, #10366 @kallewoof and #11274 @mess110). Also this one had many Concept ACKs from different contributors (not going to list them, just scroll up a bit).

So I feel like there is a threshold where something controversial can be merged if it has enough support. Another way of thinking of it, which I sometimes use to decide if a pull request should be merged: "Imagine this already was merged, but someone proposed to revert the change. Would it be worthwhile to merge?"

Back to this topic, personally I am a huge fan of the `__all__` import: https://docs.python.org/3.5/tutorial/modules.html#importing-from-a-package. I'd be in favour to take a step back and replace some of the explicit imports with an `__all__` import, where appropriate.

**Comment by kallewoof** [comment]
@laanwj 

> On one hand I'm annoyed that apparently, insisting on doing something often enough gets it merged because the people that disagree will stop bothering to complain.

Note that the previous PRs related to this were
* #11274 - closed as a duplicate of #10366 
* #10366 - closed (unexpectedly) by someone other than the author
* #9876 - closed for.. some reason. It looks like the author intended to break it into pieces before merging.

Reading through the comments, there was no one directly disagreeing with doing this. @sdaftuar had some concerns that I believe were addressed, but mostly this was a "lets not do too big changes cause we don't wanna make big changes" which is a valid reason most of the time, but IMO not here.

I.e. I don't think anyone is insisting on something repeatedly in spite of disagreement. We're all just trying to make this repo as good as it can get with what skills we have.

**Comment by laanwj** [comment]
Yes, I agree in this case.
It was probably best to merge this.
